### DISCUSSION TITLE
Roadmapping: what semver means for Shadow

### DISCUSSION THREAD

**Discussion by robgjansen** [shadow/shadow#2033] [Show and tell]
Rob will do while updating 2.0 design doc, which should be done for v2.1:
- Stability refers to command line and config file options
- i.e. within a major version, users' scripts and shadow configs should run. However the simulation is *not* guaranteed to give the same results.
- We'll document somewhere in section 3.1 "Security Guide" "Overview". We'll move the security (non) guarantee there too

Specific notes:
- Log file format is not stable
- Experimental options not stable
- File and directory *names* in the data directory *are* stable
- We won't add new files to the data dir that don't start with `<hostname>.<process name>.pid`. (We will consider removing `<hostname>` and/or changing dir layout (eg to include a sub "working_dir") in the next major version bump)
- PID numbering will stay the same (for now), allowing for "clean shutdown" with `kill`

**Comment by robgjansen** [comment]
This work is being tracked in #2041.

**Comment by stevenengler** [comment]
We had some discussion about further guarantees that we want to provide:

- default values for existing options will not change between minor versions (changing defaults requires a major version bump)
- support for a platform will not be dropped between minor versions (dropping a supported platform requires a major version bump)
- our kernel version requirement is separate from our supported platform requirement, and we should document our kernel version requirement separately
    - we believe that currently require at least kernel 5.3 (pidfd_open support)
    - we run our CI tests in docker containers for each supported platform, but that doesn't allow us to test different kernels
        - github does have a few images with different kernels, but we don't have much granularity
        - https://github.com/actions/virtual-environments
    - we may sometimes only support platforms when they use newer kernels
        - for example, ubuntu 18.04 LTS uses kernel 4.15, but if the user enables HWE ubuntu will use kernel 5.4.
        - https://ubuntu.com/about/release-cycle#ubuntu-kernel-release-cycle

**Edit:** We realized that we don't actually use pidfd_open anymore, so we don't currently have a kernel requirement.

**Comment by stevenengler** [comment]
We now have a documentation page for this: https://shadow.github.io/docs/guide/semver.html
